src/main.py

```diff
@@ -22,23 +25,22 @@ def reformat_varbinary_columns(df, varbinary_columns):
 
 
 def fetch_dune_data():
-    response = requests.get(API_URL, headers=HEADERS, timeout=10)
-    if response.status_code == 200:
-        data = response.json()
-        result = data["result"]
-        metadata = result["metadata"]
-        dtypes, varbinary_columns = {}, []
-        for name, d_type in zip(metadata["column_names"], metadata["column_types"]):
-            dtypes[name] = DUNE_TO_PG[d_type]
-            if d_type == "varbinary":
-                varbinary_columns.append(name)
-        df = pd.DataFrame(result["rows"])
-        # escape bytes
-        df = reformat_varbinary_columns(df, varbinary_columns)
-        return df, dtypes
-
-    print(f"Error fetching data: {response.status_code}")
-    return None
+    result = (
+        DuneClient(DUNE_API_KEY, performance=QUERY_ENGINE)
+        .run_query(query=QueryBase(QUERY_ID), ping_frequency=POLL_FREQUENCY)
+        .result
+    )
+
+    metadata = result.metadata
+    dtypes, varbinary_columns = {}, []
+    for name, d_type in zip(metadata.column_names, metadata.column_types):
+        dtypes[name] = DUNE_TO_PG[d_type]
```

Comment:
More of an idea than a remark - maybe each type mapping can carry an optional transformation function, so that we won't need to do the extra checks around `varbinary` below, for example.

Reply:
Yes, I like this idea a lot. Although in this loop iteration, we don't actually access the data (only the column names and types). Would be curious to see how you would propose to do this. Maybe you can create an issue for this.

Generally, all of this "transformation" (between data fetch and insert) should be implemented on it own.